### PR TITLE
[object Generator]

### PR DESCRIPTION
**Pull request by IgorNovozhilov**
25.3 Generator Objects
http://www.ecma-international.org/ecma-262/6.0/#sec-generator-objects
```
C:\Users> node
> GF = function* (){}
[Function: GF]
> GF.constructor.name
'GeneratorFunction'
> G = GF()
{}
> G + ''
'[object Generator]'
```

Fixes #

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
`Function` is specifically special-cased in our spec. Sounds like `FunctionGenerator` should be as well.

For the record, `(...args: any[]) => any` is not equivalent to `Function`. The latter is not assignable to the former. [Check out this sample](https://www.typescriptlang.org/play/#src=declare%20var%20f%3A%20Function%3B%0D%0Avar%20g%3A%20(...args%3A%20any%5B%5D)%20%3D%3E%20any%20%3D%20f%3B%20).

**Comment by IgorNovozhilov** [comment]
[then why is it working?](https://www.typescriptlang.org/play/index.html#src=declare%20var%20f%3A%20Function%3B%0D%0Avar%20g%3A%20(...args%3A%20any%5B%5D)%20%3D%3E%20any%20%3D%20f%3B%20%0D%0A%0D%0A%0D%0Ainterface%20Generator%20extends%20Iterator%3Cany%3E%20%7B%20%7D%0D%0Ainterface%20GeneratorFunction%20extends%20Function%20%7B%0D%0A%20%20%20%20(...args%3A%20any%5B%5D)%3A%20Generator%3B%0D%0A%7D%0D%0A%0D%0Adeclare%20var%20f2%3A%20GeneratorFunction%3B%0D%0Avar%20g2%3A%20(...args%3A%20any%5B%5D)%20%3D%3E%20Generator%20%3D%20f2%3B)

bug:
2017-03-23_09h42_21

not a bug:
2017-03-23_09h42_28

**Comment by mhegazy** [comment]
Sorry for the delay, a few comments:

- Generator function needs a construct signature as well
From [spec](http://www.ecma-international.org/ecma-262/6.0/#sec-generatorfunction-constructor)
> Thus the function call GeneratorFunction (…) is equivalent to the object creation expression new GeneratorFunction (…) with the same arguments.

- More importantly, this file should not be edited manually, it is auto-generated from a script in . The change should rather be done in `src/lib`

For future references, please file an issue first to track changes.

**Comment by DanielRosenwasser** [comment]
> then why is it working?

It's working as intended - like I said, `Function` isn't assignable to `(...args: any[]) => any`. Your `GeneratorFunction` definition adds an explicit signature of `(...args: any[]) => Generator`, which satisfies `(...args: any[]) => any`

**Comment by IgorNovozhilov** [comment]
This fix will be automatically assembled and updates in VSCode with these fixes? Or the lib directory is collected separately?
image

File: src/lib/es2015.generator.d.ts [hunk 1]
```diff
@@ -1,4 +1,13 @@
-interface GeneratorFunction extends Function { }
+interface Generator extends Iterator<any> { }
+
+interface GeneratorFunction {
+    /**
+     * Creates a new Generator object.
+     * @param args A list of arguments the function accepts.
+     */
+    new (...args: any[]): Generator;
+    (...args: any[]): Generator;
```
Comment (mhegazy) [inline src/lib/es2015.generator.d.ts]: nit. You will need to duplicate the comment

File: src/lib/es2015.generator.d.ts [hunk 2]
```diff
@@ -1,4 +1,13 @@
...
+interface Generator extends Iterator<any> { }
+
+interface GeneratorFunction {
+    /**
+     * Creates a new Generator object.
+     * @param args A list of arguments the function accepts.
+     */
+    new (...args: any[]): Generator;
+    (...args: any[]): Generator;
+}
```
Comment (mhegazy) [inline src/lib/es2015.generator.d.ts]: Since you are touching this, we should add `readonly length: number;`.

**Pull request merged**